R/methods_formula.R

```diff
@@ -0,0 +1,204 @@
+#' @param integrated If \code{TRUE}, the output is averaged over the
+#' parameters. Default is \code{FALSE}. For \code{proj_linpred} only.
+#' @param nv Number of variables in the submodel (the variable combination is
+#' taken from the variable selection information). If a vector with several values,
+#' then results for all specified model sizes are returned. Ignored if \code{vind} is specified.
+#' By default use the automatically suggested model size.
+#' @param draws Number of draws to return from the predictive distribution of
+#' the projection. The default is 1000.
+#' For \code{proj_predict} only.
+#' @param seed_samp An optional seed to use for drawing from the projection.
+#' For \code{proj_predict} only.
+#' @param ... Additional argument passed to \link{project} if \code{object}
+#' is an object returned by \link{varsel} or \link{cv_varsel}.
+#'
+#' @return If the prediction is done for one submodel only (\code{nv} has length one
+#' or \code{vind} is specified) and ynew is unspecified, a matrix or vector of
+#' predictions (depending on the value of \code{integrated}). If \code{ynew} is specified,
+#' returns a list with elements pred (predictions) and lpd (log predictive densities).
+#' If the predictions are done for several submodel sizes, returns a list with one element
+#' for each submodel.
+#' 
+#' @examples
+#' \donttest{
+#' ### Usage with stanreg objects
+#' fit <- stan_glm(y~x, binomial())
+#' vs <- varsel(fit)
+#' 
+#' # compute predictions with 4 variables at the training points
+#' pred <- proj_linpred(vs, xnew=x, nv = 4)
+#' pred <- proj_predict(vs, xnew=x, nv = 4)
+#' 
+#' }
+#'
+NULL
+
+# The 'helper' for proj_linpred and proj_predict, ie. does all the
+# functionality that is common to them. It essentially checks all the arguments
+# and sets them to their respective defaults and then loops over the
+# projections. For each projection, it evaluates the fun-function, which
+# calculates the linear predictor if called from proj_linpred and samples from
+# the predictive distribution if called from proj_predict.
+proj_helper_poc <- function(object, xnew, offsetnew, weightsnew, nv, seed_samp,
+                            fun, ...) {
+
+  if (is.null(offsetnew)) offsetnew <- rep(0, nrow(xnew))
+  if (is.null(weightsnew)) weightsnew <- rep(1, nrow(xnew))
+
+  if ('projection' %in% class(object) ||
+      (length(object)>0 && 'projection' %in% class(object[[1]]))) {
+    proj <- object
+  } else {
+    # reference model or varsel object obtained, so run the projection
+    proj <- project_poc(object = object, nv = nv, ...)
+  }
+
+  if (!.is_proj_list(proj)) {
+    proj <- list(proj)
+  } else {
+    # proj is not a projection object
+    if(any(sapply(proj, function(x) !('family_kl' %in% names(x)))))
+      stop(paste('proj_linpred only works with objects returned by',
+                 ' varsel, cv_varsel or project'))
+  }
+
+  ## FIXME: proj$sub_fit always exists, but does $coefficients?
```

Comment:
Not right now, but if we include projections onto gaussian processes, for instance, we will have to worry about this, because not every projection model has coefficients. Right now, as we are doing linear projections, we always have coefficients. It also largely depends on the implementation of the model, retrieving the attribute by name seems very unsafe to me, as lme4 may decide to change the name of their attributes.